### ISSUE TITLE
Rules degrade overall performance

### ISSUE THREAD

**Issue by Capoot** [junit-team/junit#1235]
I have measured test performance on a small project with ~20 test cases. When I introduced Rules, I have moved the init code into the Rule's before and after methods with cut and paste, so I expect it to be equivalent. Beforehand I have initialized my tests in the methods annotated with @Before and @After.

Avg. build execution without rules: 18.2s
Avg. build execution with rules: 30.5s
Tested with JUnit 4.12

**Comment by marcphilipp** [comment]
Can we have a look at the code?

**Comment by Tibor17** [comment]
So it means you moved the init code to TestRule and not the ClassRule,
right?
What's happening is that your tests are initializing over and over again
with every test method. Is not it this problem?
Beforehand did not you have any static flag which forced the test to be
initialized only once? Rules are created in new instance, so whatever flags
are lost.

On Wed, Jan 13, 2016 at 7:16 PM, Marc Philipp <notifications@github.com>
wrote:

> Can we have a look at the code?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/junit-team/junit/issues/1235#issuecomment-171385656>.
>



-- 
Cheers
Tibor

**Comment by Capoot** [comment]
@Tibor17 yes initializing every test method was intended and was also done before I switched to rules. I did not have any static members whatsoever.

Unfortunately I am not at liberty to post the original code. I will try to isolate a piece that can reproduce the error and share it with you later today.

Also I should note that I am using RuleChain along with the Rules

**Comment by Capoot** [comment]
Example Code at https://github.com/Capoot/rules-performance-example

I ran it with:

* Java(TM) SE Runtime Environment (build 1.8.0_45-b14)
* Apache Maven 3.3.9

Test execution with rules took about 10 sec longer on my machine (Lenovo T450s + Ubuntu 15.04) compared to execution with @Before and @After annotated method.

**Comment by kcooney** [comment]
@Capoot how are you measuring the time it takes? The difference in time should be small; somsomething weird is going on if the difference is 10 seconds.

In general, writing good micro benchmarks in Java is hard due to GC and the JIT compiler. In recommend using https://github.com/google/caliper

**Comment by Capoot** [comment]
@kcooney I just ran the tests from my IDE's UI (IntelliJ Idea), which then displays execution time. To be sure, I also ran the tests via maven from the terminal, which yields the same result. I did not write any code to measure execution time.

**Comment by marcphilipp** [comment]
@Capoot I finally found some time to look into your sample project. Thanks for providing it!

I checked out your code and noticed the rule is doing more than `setup()` and `before()` in your test that doesn't use a rule. The call of `mockServer.isRunning()` seems to be the problem. When I remove the calls in `after()` and `getMockServer()` of `MockServerRule` the tests in `MockServerWithRuleTest` run just as fast as those in `MockServerWithoutRuleTest`.
